data_structures/linked_list/circular_doubly_linked_list.c

```diff
@@ -0,0 +1,306 @@
+ *
+ * @param head              start pointer of list
+ * @return ListNode*        pointer to the list node after deleting first node
+ */
+ListNode *delete_from_tail(ListNode *head)
+{
+    if (head == NULL)
+    {
+        printf("The list is empty\n");
+        return head;
+    }
+
+    ListNode *temp1, *temp2;
+    temp1 = head;
+    temp2 = temp1->prev;
+    if (temp1 == temp2)
+    {
+        free(temp2);
+        head = NULL;
+        return head;
+    }
+    (temp2->prev)->next = temp1;
+    temp1->prev = temp2->prev;
+    free(temp2);
+    return head;
+}
+
+/**
+ * @brief The function to reverse the linked list
+ *
+ * @param head              start pointer of list
+ * @return ListNode*        pointer to the list node after reversing linked list
+ */
+ListNode *reverse_list(ListNode *head)
+{
+    if (!head)
+    {
+        printf("The list is empty\n");
+        return NULL;
+    }
+    ListNode *new_head = NULL;
+    ListNode *last = head->prev;
+    ListNode *curr = last, *prev;
+
+    // traverse list in backward direction
+    while (curr->prev != last)
+    {
+        prev = curr->prev;
+        new_head = insert_at_tail(new_head, curr->value);
+        curr = prev;
+    }
+    new_head = insert_at_tail(new_head, curr->value);
+    return new_head;
+}
+
+/**
+ * @brief The function that will return current size of list
+ *
+ * @param head      start pointer of list
+ * @return int      size of list
+ */
+int getsize(ListNode *head)
+{
+    if (!head)
+    {
+        return 0;
+    }
+    int size = 1;
+    ListNode *temp = head->next;
+    while (temp != head)
+    {
+        temp = temp->next;
+        size++;
+    }
+    return size;
+}
+
+/**
+ * @brief   Display list function
+ * @param   head   start pointer of list
+ * @returns void
+ */
+
+void display_list(ListNode *head)
+{
+    printf("\nContents of your linked list: ");
+    ListNode *temp;
+    temp = head;
+    if (head != NULL)
+    {
+        while (temp->next != head)
+        {
+            printf("%lld -> ", temp->value);
+            temp = temp->next;
+        }
+        if (temp->next == head)
+        {
+            printf("%lld", temp->value);
+        }
+    }
+    else
+    {
+        printf("The list is empty");
+    }
+    printf("\n");
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main()
+{
+    // Creating an empty list
+    ListNode *head = NULL;
+
+    // Accessing Empty list
+    assert(getsize(head) == 0);
+    printf("Current size: %d\n", getsize(head));
+    display_list(head);
+    assert(delete_from_head(head) == NULL);
+    head = delete_from_head(head);
+    display_list(head);
+    assert(delete_from_head(head) == NULL);
+    head = delete_from_tail(head);
+    display_list(head);
+
+    // Insert at head
+    head = insert_at_head(head, 3);
+    assert(insert_at_head(head, 3));
+    head = insert_at_head(head, 2);
+    head = insert_at_head(head, 1);
+    display_list(head);
```

Comment:
Your implementation lacks a function to access elements by index, which makes it hard to write tests, among others. I suggest that you implement such a function, then assert after each insertion that the element at the specified location have a correct value.